datafusion/optimizer/src/optimizer.rs

```diff
@@ -384,9 +394,16 @@ impl Optimizer {
                     // rule handles recursion itself
                     None => optimize_plan_node(new_plan, rule.as_ref(), config),
                 }
-                // verify the rule didn't change the schema
                 .and_then(|tnr| {
-                    assert_schema_is_the_same(rule.name(), &starting_schema, &tnr.data)?;
+                    // verify after each optimizer pass.
+                    assert_valid_optimization(rule.name(), &tnr.data, &starting_schema)
```

Comment:
To avoid performance regressions but still get the benefit of this check, I recommend changing this check to only run after each optimzer pass in debug mode (`#cfg(debug)`)

Reply:
Done.